doc/source/tools.rst

```diff
@@ -1,30 +1,259 @@
-  - We’ll create a helm file for our setup so that kubernetes knows how to deploy.
+Container Image
+***************
 
-**YAML**
+Container images contain the dependencies required to run your code.
+This includes **everything**, all the way down to the operating
+system itself. It also includes things like the filesystem on which
+your code runs, which might include data etc. Containers are also
+portable, meaning that you can exactly recreate the computational
+environment to run your code on almost any machine.
 
-- YAML is a generic language for structuring text, that is used in many places. 
+In Docker, images are described as layers, as in layers of dependencies.
+For example, say you want to build a container that runs scikit-learn.
+This has a dependency on Python, so you have two layers: one for
+python, and another that inherits the python layer and adds the extra
+piece of scikit-learn. Moreover, that base python layer needs an
+operating system to run on, so now you have three layers:
+ubuntu -> python -> scikit-learn. You get the idea. The beauty of this
+is that it means you can share base layers between images. This
+means that if you have many different images that all require
+ubuntu, you don't need to have many copies of ubuntu lying around.
+
+Images can be created from many things. If you're using Docker, the basic
+way to do this is with a **Dockerfile**.
+This is essentially a list of instructions that tells
+Docker how to create an image. It might tell Docker which base layers
+you want to include in an image, as well as some extra dependencies that
+you need in the image. Think of it like a recipe that tells Docker how
+to create an image.
+
+Containers
+**********
+
+You can "run" a container image, and it creates a container for you.
+A container is a particular instantiation of a container image. This means
+that it actually exists on a computer. It is a self-contained
+computational environment that is constructed according to the layers
+that are inside of the Container Image. However, because it is now
+running on the computer, it can do other useful things like talk to other
+Docker containers or communicate via the internet. 
+
+
+Kubernetes
+----------
+
+`Kubernetes <https://kubernetes.io/>`_ is a service that runs on cloud infrastructures. It provides a single point of contact with the machinery
+of your cluster deployment, and allows a user to specify the computational requirements that they need (e.g., how many machines, how many CPUs
+per machine, how much RAM).
+Then, it handles the resources on the cluster and ensures that
+these resources are always available. If something goes down, kubernetes
+will try to automatically bring it back up.
+ 
+Kubernetes can only manage the computing resources that it is
+given. This means that it generally can **not** create new resources on its
+own (with the exception of disk space).
+
+There are three main types of objects in Kubernetes:
```

Comment:
Also you have way more than 3 here :D https://www.youtube.com/watch?v=vt0Y39eMvpI